[MXNET-1333] Estimator and Fit API

roywei opened the pull request:
## Description ##
This PR introduce an Estimator class in contrib with easy fit method to help beginners with model training.
It's been developed on a branch, and we hope to merge it to contrib and get feedback for first iteration.

**Design**: https://cwiki.apache.org/confluence/display/MXNET/Gluon+Fit+API+-+Tech+Design
**JIRA epics:** https://issues.apache.org/jira/browse/MXNET-1333 
**Dev list discussion:** https://lists.apache.org/thread.html/13e3dee0fc9dd8e45b6616f97d282096a1ee67cde78a93dada295577@%3Cdev.mxnet.apache.org%3E
**Feedbacks:** currently all feedbacks are captured in [cwiki comment section](https://cwiki.apache.org/confluence/display/MXNET/Gluon+Fit+API+-+Tech+Design). We have created JIRA issues for each feedback and will continue to work on it
**Follow up PRs:**
We currently have the following PRs to address feedback, will create more and track using JIRA issue
1. [ [MXNET-1358][Fit API] Fit api tutorial ](https://github.com/apache/incubator-mxnet/pull/14462)
2. [[MXNET-1340, 1339][Fit API]Multi input/output
](https://github.com/apache/incubator-mxnet/pull/14628)



## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-1333](https://issues.apache.org/jira/browse/MXNET-1333 )
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- Estimator class for fit and evaluate
- Evenhandler class for callbacks in fit methods
- Unit tests
- Integration/nightly tests


## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

marcoabreu commented:
Would it be possible to make these tests part of the training test suite instead of introducing new jobs?

roywei commented:
@marcoabreu are you referring to training test suite [here](https://github.com/apache/incubator-mxnet/tree/master/tests/python/train)?  we want to only test this on nightly. Please point to the correct test suite if I m wrong. Thanks!

nswamy commented:
Suggest to move it to contrib as there is some feedback from Mu on the dev@. We could also gather feedback from the users to see what other changes are required.

Could you please break all the backlog items into Jira tasks and paste the master ticket to this PR ? Any contributor interested to further contribute to this could pick up those tasks.

roywei commented:
@nswamy done, all JIRA tickets has detailed description and reference to the feedback (either from cwiki or dev list discussion)

szha commented on python/mxnet/gluon/contrib/estimator/event_handler.py:
```diff
@@ -0,0 +1,346 @@
+import numpy as np
+
+
+class EventHandler(object):
```
please break this down into different event classes and use the same approach as gluon's forward hook through weakref. this has the benefit of:
- people can mix what they need into a unified handler through multi-inheritence, without the unnecessary `pass` calls.
- a handler can be detached at will.

roywei replied:
i added a [method to check if a handlers has implemented](https://github.com/apache/incubator-mxnet/pull/14629/files#diff-7f58d4d4cb6c2e6088afa89097fbb7e3R335) `train_begin` ect to avoid unnecessary `pass` calls. The time it takes should be the same as using multi-inheritence and do a bunch of `isinstance()` at the beginning. The benefit is user can override any event call without inherit multiple class.

Still looking into how to use forward hook to register different input args

Thanks!

piyushghai commented:
Thanks for your contributions @roywei. 
This is a very useful API for users and it simplifies a lot of pain points.

@mxnet-label-bot Add [Gluon, pr-awaiting-review]

roywei commented:
Thanks @szha for the feedback! I m noting the summary of offlien discussion here:
1. Currently event handlers has access to entire estimator. Estimator has to maintain different stats/states (e.g. current epoch, num of steps, metrics) and do the book keeping to ensure eventhandlers has them when they are called.
2. If user want to add an event handler that want to use some info estimator does not have. He has to add it in estimator and change the estimator code to do book keeping.
3. Separate event handlers into 6 classes (train begin class, train end class ect) instead of single eventhandler parent class. Each class maintain it's own info/state and know what args to pass when called. So estimator does not do the state management. 
4. Using the gluon forward hook approch to provide avaibilty to detach a event handler. (e.g. remove some handler after n epochs/steps)

nswamy commented:
I am not sure I understand the concerns, 

What is the problem with 1) ?

For 2), the user can create custom event Handler taking objects whatever it needs to keep track of. 
```
MyEventHandler
def __init__(self, whatever1, whatever2):
     self.whatever1 = whatever1, ...

def my_train_begin():
     # dowhatever using self.whatever1
def train_begin():
    my_train_begin()
``` 

Given that the fit API is targeted at novice users, I think 3) is going to make it unnecessarily cumbersome. 

What is the benefit of using the Forward Hook approach?

nswamy commented:
wouldn't it be easier for the user to write the training loop if they want more control instead of having the loop split 6 or more methods or hooks. 
In my opinion trying to make it more flexible would complicate the usage  and add to the users cognitive load.

roywei commented:
@nswamy @szha I have addressed the concerns on callback in this doc: https://cwiki.apache.org/confluence/display/MXNET/Callback+Design+for+Fit+Loop
and created a PR here: https://github.com/apache/incubator-mxnet/pull/14685

Please help take a look, thanks!

szha commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,363 @@
+        elif isinstance(context, list) and all([isinstance(c, Context) for c in context]):
+            context = context
+        elif not context:
+            if num_gpus() > 0:
```
since you're checking the number of GPUs, would it make sense to validate the context list too?

roywei replied:
it's checked in line 106, but looks wired, will update for better readability.

szha commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,363 @@
+                self.train_metrics = [Accuracy()]
+            self.val_metrics = []
+            for loss in self.loss:
+                self.train_metrics.append(Loss(''.join([i for i in loss.name if not i.isdigit()])))
```
```
In [6]: ''.join([i for i in mx.gluon.loss.L2Loss().name if not i.isdigit()])
Out[6]: 'lloss'
In [7]: ''.join([i for i in mx.gluon.loss.L1Loss().name if not i.isdigit()])
Out[7]: 'lloss'
```

roywei replied:
Do you think removing trailing numbers will suffice?

roywei commented:
@szha Thanks for the review, I have addressed the comments here: https://github.com/apache/incubator-mxnet/pull/14849, let me know if there is any other issue, I will adress it together in one PR.

roywei commented on ci/docker/runtime_functions.sh:
```diff
@@ -1296,6 +1296,16 @@ nightly_scala_demo_test_cpu() {
+    python test_estimator_cnn.py --type gpu
+    python test_sentiment_rnn.py --type gpu
+    python test_estimator_cnn.py --type cpu
+    python test_sentiment_rnn.py --type cpu
```
we are running python scripts instead of nosetests, but both will throw error 
I'm combining these 4 tests, although running them separately locally take less than 5 min, but creating a separate job and run on CI takes long time. So I'm adding total 1 additional job to run.

roywei commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,376 @@
+            if gpus > 0:
+                # only use 1 GPU by default
+                if gpus > 1:
+                    warnings.warn("You have multiple GPUs, gpu(0) will be used by default."
```
I'm using 1 gpu by default and give warning if multiple gpu detected. borrowed behavior from keras, as new/beginner users always try small datasets.

roywei commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,376 @@
+        return True
+
+    def _get_data_and_label(self, batch, ctx):
+        data = batch[0]
```
assuming 1 data 1 label for now, multi input/output will be supported by follow up PR once we can finalize on the estimator and callback design. was WIP in https://github.com/apache/incubator-mxnet/pull/14628, will reopen once this merged.

roywei commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,376 @@
+            self.val_metrics = []
+            for loss in self.loss:
+                # remove trailing numbers from loss name to avoid confusion
+                self.train_metrics.append(Loss(loss.name.rstrip('1234567890')))
```
bydefault, loss.name will have node number in name, I'm removing trailing numbers.
```
>>>loss=mx.gluon.loss.L2Loss()
>>> loss.name
'l2loss0'
```

roywei commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,376 @@
+
+                with autograd.record():
+                    pred = [self.net(x) for x in data]
+                    loss = [self.loss[0](y_hat, y) for y_hat, y in zip(pred, label)]
```
same as above, support 1 loss for now, will reopen https://github.com/apache/incubator-mxnet/pull/14628 once this is finalized.

roywei commented on python/mxnet/gluon/trainer.py:
```diff
@@ -255,6 +255,13 @@ def learning_rate(self):
             return self._optimizer.learning_rate
 
+    @property
+    def optimizer(self):
```
I m exposing optimizer through trainer, as estimator only has access to trainer. For now it's only used for logging and type and name. Let me know if I should remove this.

roywei commented on tests/nightly/estimator/test_estimator_cnn.py:
```diff
@@ -0,0 +1,151 @@
+    '''
+    Test estimator by doing one pass over each model with synthetic data
+    '''
+    models = ['resnet18_v1',
```
on CPU it's testing different models with synthetic data. list of models should increase as more models are supported. maybe `test_estimator_cpu()` should take models as args

roywei commented on tests/nightly/estimator/test_estimator_cnn.py:
```diff
@@ -0,0 +1,151 @@
+    '''
+    Test estimator by training resnet18_v1 for 5 epochs on MNIST and verify accuracy
+    '''
+    model_name = 'resnet18_v1'
```
on gpu it's choosing one fixed model and running on actual data and check for acc.

roywei commented on tests/nightly/estimator/test_sentiment_rnn.py:
```diff
@@ -0,0 +1,276 @@
+    '''
+    Test estimator by doing one pass over each model with synthetic data
+    '''
+    models = ['TextCNN', 'BiRNN']
```
save behavior as cnn test

eric-haibin-lin commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,376 @@
+
+
+    Parameters
+    ----------
```
`net` didn't appear in the documentation.

roywei replied:
added

eric-haibin-lin commented on python/mxnet/gluon/contrib/estimator/event_handler.py:
```diff
@@ -0,0 +1,459 @@
+        self._save_checkpoint(estimator.net, "Epoch", self.num_epochs)
+        self.num_epochs += 1
+
+    def _save_checkpoint(self, net, period_name, period_value):
```
trainer should also be part of the interface. trainer states should be saved, too

roywei replied:
tracked in https://issues.apache.org/jira/browse/MXNET-1373

eric-haibin-lin commented on python/mxnet/gluon/contrib/estimator/event_handler.py:
```diff
@@ -0,0 +1,459 @@
+            self.batch_index = 0
+
+
+class CheckpointHandler(BatchEnd, EpochEnd):
```
I don't see how an estimator can use the produced checkpoints and resume training from there. Is there any intention to cover that usage? With the tensorflow estimator api https://www.tensorflow.org/guide/checkpoints, it's very easy - as long as the `model_dir` argument is set. User don't need to change any code. 
To do that we also need to add logic to save trainer state and resume model parameter/trainer states.

roywei replied:
as we discussed offline before, we need to decide to save a few more information (serialize train history) in addition to params and trainer state. WIll need a bit of design, tacked https://issues.apache.org/jira/browse/MXNET-1373, will open another PR to add this feature after we can finalize general event handler design.

roywei replied:
added functionality to save symbol, param, and trainer states in model_dir, only keep a maximum of check points (default 5).  resuming training can be done mannually now, need to design for automatical resuming later.

pinaraws commented on python/mxnet/gluon/contrib/estimator/event_handler.py:
```diff
@@ -0,0 +1,459 @@
+        optimizer = trainer.optimizer.__class__.__name__
+        lr = trainer.learning_rate
+        self.logger.info("Training begin: using optimizer %s "
+                         "with current learning rate %.4f ",
```
print learning rate at every epoch

roywei replied:
added

pinaraws commented on python/mxnet/gluon/contrib/estimator/event_handler.py:
```diff
@@ -0,0 +1,459 @@
+        self.num_batches += 1
+
+    def epoch_end(self, estimator, *args, **kwargs):
+        self._save_checkpoint(estimator.net, "Epoch", self.num_epochs)
```
this will create a lot of files with unique names, should we also allow to save the last only in the same filename?

roywei replied:
I'm using save parameter only, epcoh/batch number is not saved into file name.  num_epoch and num_batch is passed only to check if the saving frequency has reached.

pinaraws commented on python/mxnet/gluon/contrib/estimator/event_handler.py:
```diff
@@ -0,0 +1,459 @@
+                self.best = np.Inf
+
+    def batch_end(self, estimator, *args, **kwargs):
+        self._save_checkpoint(estimator.net, "Batch", self.num_batches)
```
is this typical? maybe this should be optional

roywei replied:
added check, default does not save based on batch

pinaraws commented on python/mxnet/gluon/contrib/estimator/event_handler.py:
```diff
@@ -0,0 +1,459 @@
+class CheckpointHandler(BatchEnd, EpochEnd):
+    """Save the model after every epoch.
+
+    :py:class:`CheckpointHandler` save the network parameters every epoch
```
how about adding frequency? default to 1

roywei replied:
added documentation for `epoch_period` default 1 and `batch_period` default None. So by default checkpoint every 1 epoch and does not save based on batch/steps

roywei commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,376 @@
+    def _check_loss(self, loss):
+        if isinstance(loss, gluon.loss.Loss):
+            loss = [loss]
+        elif isinstance(loss, list) or all([isinstance(l, gluon.loss.Loss) for l in loss]):
```
fixed

roywei commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,376 @@
+        if isinstance(metrics, EvalMetric):
+            metrics = [metrics]
+        else:
+            metrics = metrics or []
```
yes, we can still train only based on loss value.  by default, we will wrap loss with metric.Loss() to record it's value. user does not need to specify metrics. Tested in unit tests

roywei commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,376 @@
+    def _initialize(self, initializer):
+        # initialize the network
+        if initializer:
+            if self._is_initialized():
```
removed force init logic,  now if the net is partially initialized, only the uninitialized layers will be initialized.

roywei commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,376 @@
+    def _get_data_and_label(self, batch, ctx):
+        data = batch[0]
+        label = batch[1]
+        data = gluon.utils.split_and_load(data, ctx_list=ctx, batch_axis=0)
```
exposed batch_axis, and put it as default 0, added docuemntation

roywei commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,376 @@
+    def evaluate(self,
+                 val_data,
+                 val_metrics):
+        """Evaluate model on validation data
```
only support dataloader, added check in both evaluated and fit, provided message on converting DataIter and ndarray to data loader. We can add a predict function later to support one ndarray directly. Updated doc

roywei commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,376 @@
+            validation data with data and labels
+        epochs : int, default 1
+            number of epochs to iterate on the training data.
+        batch_size : int
```
removed

roywei commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,376 @@
+        batch_end, epoch_end, train_end = self._categorize_handlers(event_handlers)
+
+        # only pass a weak reference to all event handlers
+        estimator_ref = weakref.proxy(self)
```
removed

roywei commented on python/mxnet/gluon/contrib/estimator/estimator.py:
```diff
@@ -0,0 +1,376 @@
+                handler.epoch_begin(estimator_ref)
+
+            for i, batch in enumerate(train_data):
+                if not isinstance(train_data, gluon.data.DataLoader):
```
done

roywei commented on python/mxnet/gluon/contrib/estimator/event_handler.py:
```diff
@@ -0,0 +1,459 @@
+                             % verbose)
+        self.verbose = verbose
+        # save logger to file only if file name or location is specified
+        if file_name or file_location:
```
added file mode option, default is 'a' which will append, updated documentation

roywei commented on python/mxnet/gluon/contrib/estimator/event_handler.py:
```diff
@@ -0,0 +1,459 @@
+    save_best_only: bool
+        if True, only save the parameters if monitored value improved
+    mode: str, default 'auto'
+        one of {auto, min, max}, if `save_best_only=True`, the comparison to make
```
done, added in doc string, warning if auto is used, added logging message if auto has determined to use np.less or np.greater

roywei commented on python/mxnet/gluon/contrib/estimator/event_handler.py:
```diff
@@ -0,0 +1,459 @@
+            self.best = -np.Inf
+        else:
+            # use greater for accuracy and less otherwise
+            if 'acc' in self.monitor.get()[0].lower():
```
added f1

roywei commented:
@eric-haibin-lin @pinaraws @szha Thanks for the review, address comments here: https://github.com/apache/incubator-mxnet/pull/14885/files

eric-haibin-lin commented:
My comments are addressed. Great work!!

roywei commented:
@szha @eric-haibin-lin  CI finally passed, [validation/miscellaneous  job](http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Fmiscellaneous/detail/PR-14629/14/pipeline) status returned not correctly (passed instead of pending). Could you help merge if looks good?

szha commented:
Nice work! Great job upholding the quality even at the cost of several iterations. Well done!
